### PR TITLE
New Marketing Pref component

### LINKED ISSUE #378: New Marketing Preferences component/s

### PR DESCRIPTION
**Pull request by AndyEPhipps**
Type: patch

Fixes #378 

## Changes proposed in this PR

- Briefly describe the changes in this PR
- Do you have some test asserting this behavior?

### CODE REVIEW

File: src/components/Organisms/MarketingPreferencesDS/_MarketingPreferencesDS.js [hunk 1]
```diff
@@ -0,0 +1,259 @@
...
+                ref={allRefs.town}
+              />
+              <TextInput
+                placeholder=""
+                fieldName={prefixName('postcode')}
+                label="Postcode"
+                onTextFieldChange={onTextFieldChange}
+                fieldValuesFromParent={fieldVals}
+                isRequired
+                ref={allRefs.postcode}
```
Comment (AndyEPhipps) [inline src/components/Organisms/MarketingPreferencesDS/_MarketingPreferencesDS.js:211]: pooass

**Comment by tomisaacmarshall** [comment]
@AndyEPhipps do you have a link to the designs for this?

**Comment by AndyEPhipps** [comment]
> @AndyEPhipps do you have a link to the designs for this?

Ah yeah, sorry- here you go:

https://www.figma.com/file/MtusIm9S0M6Tj76AWVGcYe/CR-BAU-MarketingPrefs?node-id=166%3A99

I'd done a bit of a collab updates on this with Ruby, so not 100% sure if that's been updated. I'm gonna show her the build now too

File: package.json [hunk 2]
```diff
@@ -22,7 +22,9 @@
   },
   "dependencies": {
     "@babel/cli": "^7.10.4",
+    "@hookform/resolvers": "^1.3.4",
     "axios": "^0.19.2",
+    "formik": "^2.2.6",
```
Comment (gusliedke) [inline package.json:27]: I think we can remove `formik` right?

Comment (AndyEPhipps) [reply package.json:27]: Ooh yeah, good spot

File: package.json [hunk 3]
```diff
@@ -37,7 +39,10 @@
     "react-test-renderer": "^16.8.6",
     "react-uid": "^2.2.0",
     "styled-components": "^5.1.1",
-    "youtube-player": "^5.5.2"
+    "use-deep-compare-effect": "^1.6.1",
```
Comment (AndyEPhipps) [inline package.json:42]: Will remove this too

File: src/components/Organisms/MarketingPreferencesDS/MarketingPreferencesDS.style.js [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+  align-items: center;
+`;
+
+const CheckInput = styled.input`
+  font-size: ${({ theme }) => theme.fontSize('sm')};
+  display: block;
+  box-sizing: border-box;
+  opacity: 0;
+  position: absolute;
+  //width: 100%;
```
Comment (gusliedke) [inline src/components/Organisms/MarketingPreferencesDS/MarketingPreferencesDS.style.js:101]: unnecessary?

File: src/components/Organisms/MarketingPreferencesDS/MarketingPreferencesDS.style.js [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+  :checked + span {
+    background: url(${checkBoxIcon}) no-repeat center;
+    background-color: ${({ theme }) => theme.color('red')};
+    background-size: contain;
+  }
+  :focus + span {
+    border-color: ${({ theme }) => theme.color('grey_for_forms')};
+    border-width: 3px;
+  }
+  :focus + span {
```
Comment (gusliedke) [inline src/components/Organisms/MarketingPreferencesDS/MarketingPreferencesDS.style.js:125]: duplicated lines

**Comment by AndyEPhipps** [comment]
NB: I tried and failed to set up a Jest test for this; not able to import the required dependencies within the 'plain JS' context

File: src/components/Organisms/MarketingPreferencesDS/MarketingPreferencesDS.style.js (hunk 5 above)
Comment (AndyEPhipps) [reply src/components/Organisms/MarketingPreferencesDS/MarketingPreferencesDS.style.js:125]: Odd. Wow, I really should've given my code a proper once-over before review, sorry guys.

File: src/components/Organisms/MarketingPreferencesDS/MarketingPreferencesDS.style.js (hunk 4 above)
Comment (AndyEPhipps) [reply src/components/Organisms/MarketingPreferencesDS/MarketingPreferencesDS.style.js:101]: Aha, this was addressing a bug that has since been fixed, will check if they need to go back in or not now

File: src/components/Organisms/MarketingPreferencesDS/MarketingPreferencesDSForm.js [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+const mpValidation = buildValidationSchema();
+const initialValues = setInitialValues();
+
+const {
+  mpValidationSchema,
+  mpValidationOptions
+} = mpValidation;
+
+/*
+* This component exists purely to show the Marketing Preferences component functioning
```
Comment (gusliedke) [inline src/components/Organisms/MarketingPreferencesDS/MarketingPreferencesDSForm.js:20]: I think you can do this in `MarketingPreferencesDS.md` file.

File: src/components/Organisms/MarketingPreferencesDS/assets/CR_@.svg [hunk 7]
```diff
@@ -0,0 +1,21 @@
+<?xml version="1.0" encoding="utf-8"?>
+<!-- Generator: Adobe Illustrator 24.3.0, SVG Export Plug-In . SVG Version: 6.00 Build 0)  -->
```
Comment (gusliedke) [inline src/components/Organisms/MarketingPreferencesDS/assets/CR_@.svg:2]: I use `svgo` as a global package to optimize svgs.. https://www.npmjs.com/package/svgo

**Comment by gusliedke** [comment]
Looks very nice! Maybe worth to create a CR env using it to check if the changes on form fields haven't affect other forms.

**Comment by AndyEPhipps** [comment]
> Looks very nice! Maybe worth to create a CR env using it to check if the changes on form fields haven't affect other forms.

Thanks man, I'm just finishing up the integration with the Prize platform form now- I think I've been specific with the field name=spacing to avoid any weird overlaps; it seems to be working fine :)

**Review by tomisaacmarshall — CHANGES_REQUESTED** [review]
Looks good man, just a few comments, in particular could do with a bit more info / examples re the config options

File: src/components/Organisms/MarketingPreferencesDS/_TextInput.js [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+    type: fieldType,
+    label,
+    placeholder: label,
+    errorMsg: errors && errors[fieldName]
+      && errors[fieldName].message,
+    isRequired,
+    'aria-required': isRequired,
+    labelProps: {
+      className: errors && errors[fieldName]
+      && errors[fieldName].message ? 'error-message' : ''
```
Comment (tomisaacmarshall) [inline src/components/Organisms/MarketingPreferencesDS/_TextInput.js:23]: I can't seem to see any styling associated with this class, is it needed?

File: src/components/Organisms/MarketingPreferencesDS/_MarketingPrefsConfig.js [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+  return updatedValues;
+};
+
+const buildValidationSchema = overrideOptions => {
+  const defaultOptions = {
+    mp_permissionEmail: {
+      /**
+        * As per react-hook-form's validation requirements, sets the fields' required attribute
+        * for each checkbox option (Yes & No), a non-required field isn't rendered or included
+        * in the validation. As ERP *currently* needs values to formalise a user's opt-out,
```
Comment (tomisaacmarshall) [inline src/components/Organisms/MarketingPreferencesDS/_MarketingPrefsConfig.js:34]: Minor point but I don't think this restriction is to do with the ERP, so would maybe just say 'back end'

File: src/components/Organisms/MarketingPreferencesDS/_MarketingPrefsConfig.js [hunk 10]
```diff
@@ -0,0 +1,144 @@
...
+    mp_permissionSMS: []
+  };
+
+  // Override with any custom options supplied
+  const updatedValues = merge(defaultValues, overrideValues);
+
+  return updatedValues;
+};
+
+const buildValidationSchema = overrideOptions => {
```
Comment (tomisaacmarshall) [inline src/components/Organisms/MarketingPreferencesDS/_MarketingPrefsConfig.js:28]: I somehow completely missed this file on first look, & it helps a lot with filling in some of the blanks re config!

File: src/components/Organisms/MarketingPreferencesDS/_MarketingPrefsConfig.js [hunk 11]
```diff
@@ -0,0 +1,144 @@
...
+
+    mp_mobile: yup.string()
+      .when('mp_permissionSMS', {
+        is: val => (mpValidationOptions.mp_permissionSMS[val]),
+        then: yup.string().required('Please enter your mobile number')
+      })
+      .when('mp_permissionSMS', {
+        is: val => (!(mpValidationOptions.mp_permissionSMS.disableOption)
+        && mpValidationOptions.mp_permissionSMS[val]),
+        then: yup.string().phone('GB', false, 'Please enter a valid mobile number')
```
Comment (tomisaacmarshall) [inline src/components/Organisms/MarketingPreferencesDS/_MarketingPrefsConfig.js:82]: Could we just add `.required('Please enter your mobile number')` here and have only one `when` condition?

File: src/components/Atoms/Input/Input.js [hunk 12]
```diff
@@ -92,6 +95,7 @@ const Input = React.forwardRef(
           aria-describedby={hasAria ? id : undefined}
           ref={ref}
           prefixLength={prefix.length}
+          required
```
[the remark this answers is not in the document]
Comment (AndyEPhipps) [reply src/components/Atoms/Input/Input.js:98]: Whoops, didn't pass in the prop for some reason

**Comment by tomisaacmarshall** [comment]
> > @AndyEPhipps do you have a link to the designs for this?
> 
> Ah yeah, sorry- here you go:
> 
> https://www.figma.com/file/MtusIm9S0M6Tj76AWVGcYe/CR-BAU-MarketingPrefs?node-id=166%3A99
> 
> I'd done a bit of a collab updates on this with Ruby, so not 100% sure if that's been updated. I'm gonna show her the build now too

Don't seem to have permish for this but doesn't matter!

File: package.json (hunk 2 above)
[the remark this answers is not in the document]
Comment (AndyEPhipps) [reply package.json]: Gus caught it, now removed :)

File: src/components/Atoms/Input/Input.js (hunk 12 above)
[the remark this answers is not in the document]
Comment (tomisaacmarshall) [reply src/components/Atoms/Input/Input.js:98]: Also wonder if we want to see the browser validation in addition to our own?

File: src/components/Atoms/Input/Input.js [hunk 13]
```diff
@@ -129,7 +137,8 @@ Input.defaultProps = {
   errorMsg: '',
   labelProps: {},
   className: '',
-  prefix: ''
+  prefix: '',
+  isRequired: false
```
[the remark this answers is not in the document]
Comment (AndyEPhipps) [reply src/components/Atoms/Input/Input.js:141]: Hmmmm yeah, I think you're totally right; will update

File: src/components/Organisms/MarketingPreferencesDS/_HelperFunctions.js [hunk 14]
```diff
@@ -0,0 +1,8 @@
+const associatedFields = {
```
[the remark this answers is not in the document]
Comment (AndyEPhipps) [reply src/components/Organisms/MarketingPreferencesDS/_HelperFunctions.js:1]: I guess I was expecting more functions to join the party

File: src/components/Atoms/ErrorText/ErrorText.js [hunk 15]
```diff
@@ -7,7 +7,14 @@ import alertIcon from '../Input/assets/error-alert-icon.png';
...
 const ErrorIcon = styled.span`
   display: inline-block;
-  margin-top: ${spacing('sm')};
+  padding: ${spacing('m')} ${spacing('md')} ${spacing('md')};
+  width: 100%;
+  z-index: 1;
+  margin-top: -10px;
+  background-color: ${({ theme }) => theme.color('error_pink')};
+  border-radius: ${spacing('sm')};
+  font-weight: normal;
```
[the remark this answers is not in the document]
Comment (AndyEPhipps) [reply src/components/Atoms/ErrorText/ErrorText.js]: I've not seen it look like that; likely my browser doing something yours, probs a missing position: relative or something

File: src/components/Atoms/Label/Label.js [hunk 16]
```diff
@@ -9,25 +9,39 @@ import hideVisually from '../../../theme/shared/hideVisually';
...
-  color: ${({ theme }) => theme.color('grey_label')};
+  color: ${({ theme }) => theme.color('grey_label')}; 
 `;
 const VisibleText = styled(Text).attrs({ weight: 'bold' })`
-  margin-bottom: ${spacing('sm')}
+  margin-bottom: ${spacing('sm')};
+
+  ${({ isRequired }) => isRequired === false && `
+  :after {
+    position: absolute;
```
[the remark this answers is not in the document]
Comment (AndyEPhipps) [reply src/components/Atoms/Label/Label.js:19]: The label already has it

File: src/components/Atoms/ErrorText/ErrorText.js (hunk 15 above)
[the remark this answers is not in the document]
Comment (tomisaacmarshall) [reply src/components/Atoms/ErrorText/ErrorText.js]: It was on the normal input rather than marketing prefs component: https://deploy-preview-380--cr-component-library.netlify.app/#!/Input/1

[the remark this answers is not in the document]
Comment (tomisaacmarshall) [reply src/components/Atoms/ErrorText/ErrorText.js]: (on Chrome)

File: src/components/Atoms/Label/Label.js (hunk 16 above)
[the remark this answers is not in the document]
Comment (tomisaacmarshall) [reply src/components/Atoms/Label/Label.js:19]: That `FormField` component is specific to the marketing prefs, whereas this `Label` component is used across all forms, so its styling needs to stand alone

File: src/components/Atoms/Input/Input.js (hunk 12 above)
[the remark this answers is not in the document]
Comment (AndyEPhipps) [reply src/components/Atoms/Input/Input.js]: Browser validation was overriding the react-form-hook validation (hence using NoValidate on the form itself), but I feel it's better UX and accessibility to have it marked as required when it is.

File: src/components/Atoms/Label/Label.js (hunk 16 above)
[the remark this answers is not in the document]
Comment (AndyEPhipps) [reply src/components/Atoms/Label/Label.js:19]: So being so right all the time

File: src/components/Atoms/Input/Input.js (hunk 12 above)
[the remark this answers is not in the document]
Comment (tomisaacmarshall) [reply src/components/Atoms/Input/Input.js]: Ah, I wasn't familiar with this `noValidate` prop, thanks!

If `required` improves accessibility, definitely a case to use it, however, if browser validation indeed overrides react-hook-form validation, it would be an issue to add it as part of this PR - since this component is already being used in production by other forms which _don't_ have the `noValidate` prop

File: src/components/Atoms/ErrorText/ErrorText.js (hunk 15 above)
[the remark this answers is not in the document]
Comment (AndyEPhipps) [reply src/components/Atoms/ErrorText/ErrorText.js]: Aha right; looks like it was an existing problem, but the new updated bg colour makes it more obviously. Will bung a width on it

File: src/components/Organisms/MarketingPreferencesDS/MarketingPreferencesDSForm.js [hunk 17]
```diff
@@ -0,0 +1,53 @@
...
+* within the Component Library; applications are to provide their own react-hook-form form
+* and validation, based on these
+*/
```
[the remark this answers is not in the document]
Comment (AndyEPhipps) [reply src/components/Organisms/MarketingPreferencesDS/MarketingPreferencesDSForm.js]: Since moved!

File: src/components/Organisms/MarketingPreferencesDS/MarketingPreferencesDSForm.js [hunk 18]
```diff
@@ -0,0 +1,53 @@
...
+    resolver: yupResolver(mpValidationSchema)
+  });
+
+  const { handleSubmit } = formMethods;
+
+  return (
+    <>
+      <FormProvider {...formMethods}>
+        <form onSubmit={handleSubmit(customSubmitHandler)} noValidate>
+          <input type="submit" />
```
[the remark this answers is not in the document]
Comment (AndyEPhipps) [reply src/components/Organisms/MarketingPreferencesDS/MarketingPreferencesDSForm.js:38]: I did have examples of this initially, but figured it made it hard to follow, although that _was_ when I was using Formik, which was already way messier by design.

I'll see what's possible here; not sure if we'll run into issues of having field attribute duplicates on one page if we're showing multiple instances

File: src/components/Organisms/MarketingPreferencesDS/_MarketingPreferencesDS.js [hunk 19]
```diff
@@ -0,0 +1,229 @@
...
+} from './MarketingPreferencesDS.style';
+
+import {
+  setInitialValues, buildValidationSchema
+} from './_MarketingPrefsConfig';
+
+const MarketingPreferencesDS = ({
+  copyTop,
+  copyBottom,
+  validationOptions
```
[the remark this answers is not in the document]
Comment (AndyEPhipps) [reply src/components/Organisms/MarketingPreferencesDS/_MarketingPreferencesDS.js:22]: The defaults are defined within `MarketingPrefsConfig.js` in the `buildValidationSchema` (as the Yup rules are based on the config); can maybe just signpost to that, as there's a better explanation there

File: src/components/Organisms/MarketingPreferencesDS/MarketingPreferencesDSForm.js (hunk 18 above)
[the remark this answers is not in the document]
Comment (tomisaacmarshall) [reply src/components/Organisms/MarketingPreferencesDS/MarketingPreferencesDSForm.js:38]: Ah, yeah, if they are all wrapped in distinct `<form>` tags I _think_ it'll be fine

File: src/components/Organisms/MarketingPreferencesDS/_TextInput.js (hunk 8 above)
Comment (AndyEPhipps) [reply src/components/Organisms/MarketingPreferencesDS/_TextInput.js:23]: I think this might be some vestigial code from a previous design choice, will strip it owtttt

File: src/components/Organisms/MarketingPreferencesDS/_MarketingPrefsConfig.js (hunk 9 above)
Comment (AndyEPhipps) [reply src/components/Organisms/MarketingPreferencesDS/_MarketingPrefsConfig.js:34]: > > > @AndyEPhipps do you have a link to the designs for this?
> > 
> > 
> > Ah yeah, sorry- here you go:
> > https://www.figma.com/file/MtusIm9S0M6Tj76AWVGcYe/CR-BAU-MarketingPrefs?node-id=166%3A99
> > I'd done a bit of a collab updates on this with Ruby, so not 100% sure if that's been updated. I'm gonna show her the build now too
> 
> Don't seem to have permish for this but doesn't matter!

I'm making a few tweaks with Ruby, i'll get her official stamp of approval eventually  😉

File: src/components/Organisms/MarketingPreferencesDS/_MarketingPrefsConfig.js (hunk 11 above)
Comment (AndyEPhipps) [reply src/components/Organisms/MarketingPreferencesDS/_MarketingPrefsConfig.js:82]: Ah yeah, nice; I just removed an additional check that Formik needed, but now there's a bit of duplication; will simplify where I can

File: src/components/Atoms/Input/Input.js (hunk 12 above)
[the remark this answers is not in the document]
Comment (AndyEPhipps) [reply src/components/Atoms/Input/Input.js]: Hmm yeah, that sounds like a point. Okay, I'll remove this for now and keep any 'Optional' labelling local to the MPs

[the remark this answers is not in the document]
Comment (tomisaacmarshall) [reply src/components/Atoms/Input/Input.js]: It's nice have the Optional label available on the general Input component - as long as it's optional itself rather than the default 

Once contentful gets updated with latest CL, I'll add the label to any fundraiser form inputs that need it

[the remark this answers is not in the document]
Comment (AndyEPhipps) [reply src/components/Atoms/Input/Input.js]: Will revisit after lunch

File: src/components/Atoms/Input/Input.js [hunk 20]
```diff
@@ -92,6 +95,7 @@ const Input = React.forwardRef(
           aria-describedby={hasAria ? id : undefined}
           ref={ref}
           prefixLength={prefix.length}
+          required={optional === false}
```
Comment (tomisaacmarshall) [inline src/components/Atoms/Input/Input.js:98]: Could move this line above `{...rest}` so that `<Input required>` still works?

**Review by tomisaacmarshall — APPROVED** [review]
Looks good! One minor suggestion at your discretion. Also it'd be safe to remove the old MarketingPrefs component now if you wanted

If you add me on the PR when updating CL on contentful I can check the fundraising form

File: src/components/Atoms/Input/Input.js (hunk 20 above)
Comment (AndyEPhipps) [reply src/components/Atoms/Input/Input.js:98]: Ooh yeah, very good call

Comment (tomisaacmarshall) [reply src/components/Atoms/Input/Input.js:97]: Yeah makes sense
